### PR TITLE
[PSEH/x64] Fix our GCC-SEH-plugin hack

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

The change is needed, because the previous version emitted multiple ".seh_handlerdata" blocks and GAS isn't able to merge them into one, but will instead only include the first one. This is fixed by generating "asm defines" in the code and pass the line number to the "REACTOS seh" pragma, which is then used to create all handler data by referring to the predefined symbols, that include the line number. Also the EstablisherFrame parameter passed to the filter/finally function is the original stack pointer, not the frame pointer! Take this into account by passing __builtin_frame_address(0) from the filter context to the global wrapper function, which includes the frame-offset, and use that to recalculate the frame-pointer from the passed in stack pointer.

## Proposed changes
- Fix problem with multiple .seh_handlerdata blocks being ignored by creating a full one from the plugin
- Fix frame pointer offset when calling the exception filter / finally function

### CODE REVIEW

File: sdk/tools/gcc_plugin_seh/main.cpp [hunk 1]
```diff
@@ -31,23 +32,28 @@ int
...
     seh_function(struct function* fun)
         : unwind(false)
         , except(false)
         , count(0)
     {
         /* Reserve space for our header statement */
-        char buf[256];
-        memset(buf, 0, sizeof(buf));
-        asm_header_text = build_string(sizeof(buf), buf);
+        asm_header_text = build_string(20000, "");
```
Comment (HBelusca) [inline sdk/tools/gcc_plugin_seh/main.cpp:56]: I wonder whether this "20000" could be made a constexpr or a define somewhere in this file.

**Comment by HBelusca** [comment]
Looks like gcc x64 is pissed off.

File: sdk/tools/gcc_plugin_seh/main.cpp [hunk 2]
```diff
@@ -142,12 +162,25 @@ finish_seh_function(void* event_data, void* UNUSED user_data)
...
+        else
+            asm_str << "\n\t.long 0"; /* No unwind handler */
+    }
+    asm_str << "\n\t.seh_code\n";
 
     strncpy(const_cast<char*>(TREE_STRING_POINTER(seh_fun->asm_header_text)),
             asm_str.str().c_str(),
             TREE_STRING_LENGTH(seh_fun->asm_header_text));
 
+    //fprintf(stderr, "ASM: %s\n", asm_str.str().c_str());
```
Comment (tkreuzer) [inline sdk/tools/gcc_plugin_seh/main.cpp:182]: Note to self: make this a trace macro.

Comment (tkreuzer) [reply sdk/tools/gcc_plugin_seh/main.cpp]: done

**Comment by binarymaster** [comment]
@tkreuzer it seems GCC amd64 is broken on master:

image

**Comment by binarymaster** [comment]
```sh
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x3c): undefined reference to `__seh2$$begin_try__121'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x40): undefined reference to `__seh2$$end_try__121'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x44): undefined reference to `__seh2$$filter__121'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x48): undefined reference to `__seh2$$begin_except__121'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x4c): undefined reference to `__seh2$$begin_try__126'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x50): undefined reference to `__seh2$$end_try__126'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x54): undefined reference to `__seh2$$filter__126'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x58): undefined reference to `__seh2$$begin_except__126'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x5c): undefined reference to `__seh2$$begin_try__143'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x60): undefined reference to `__seh2$$end_try__143'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x64): undefined reference to `__seh2$$filter__143'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x68): undefined reference to `__seh2$$begin_except__143'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x6c): undefined reference to `__seh2$$begin_try__147'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x70): undefined reference to `__seh2$$end_try__147'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x74): undefined reference to `__seh2$$filter__147'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x78): undefined reference to `__seh2$$begin_except__147'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x7c): undefined reference to `__seh2$$begin_try__151'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x80): undefined reference to `__seh2$$end_try__151'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x84): undefined reference to `__seh2$$filter__151'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x88): undefined reference to `__seh2$$begin_except__151'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x8c): undefined reference to `__seh2$$begin_try__155'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x90): undefined reference to `__seh2$$end_try__155'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x94): undefined reference to `__seh2$$filter__155'
/home/runner/work/reactos/reactos/RosBE-CI/amd64/lib/gcc/x86_64-w64-mingw32/8.4.0/../../../../x86_64-w64-mingw32/bin/ld: modules/rostests/kmtests/CMakeFiles/kmtest_drv.dir/rtl/RtlException.c.obj:RtlException.c:(.xdata+0x98): undefined reference to `__seh2$$begin_except__155'
collect2: error: ld returned 1 exit status
```

**Pull request merged**